cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/AuthorizationArtifactRepository.java

```diff
@@ -0,0 +1,345 @@
+      return artifacts;
+    }
+    final Principal principal = authenticationContext.getPrincipal();
+    return Lists.newArrayList(
+      Iterables.filter(artifacts, new com.google.common.base.Predicate<ArtifactDetail>() {
+        @Override
+        public boolean apply(ArtifactDetail artifactDetail) {
+          co.cask.cdap.api.artifact.ArtifactId artifactId = artifactDetail.getDescriptor().getArtifactId();
+          co.cask.cdap.proto.id.ArtifactId newArtifact =
+            new NamespaceId(range.getNamespace()).artifact(artifactId.getName(),
+                                                           artifactId.getVersion().getVersion());
+          try {
+            return !authorizationEnforcer.isVisible(Collections.singleton(newArtifact), principal).isEmpty();
+          } catch (Exception e) {
+            LOG.warn("Error checking visibility for artifact: {}", artifactId);
+            return false;
+          }
+        }
+      })
+    );
+  }
+
+  @Override
+  public List<ApplicationClassSummary> getApplicationClasses(NamespaceId namespace,
+                                                             boolean includeSystem) throws IOException {
+    return delegate.getApplicationClasses(namespace, includeSystem);
+  }
+
+  @Override
+  public List<ApplicationClassInfo> getApplicationClasses(NamespaceId namespace, String className) throws IOException {
+    return delegate.getApplicationClasses(namespace, className);
+  }
+
+  @Override
+  public SortedMap<ArtifactDescriptor, Set<PluginClass>> getPlugins(
+    NamespaceId namespace, Id.Artifact artifactId) throws IOException, ArtifactNotFoundException {
+    return delegate.getPlugins(namespace, artifactId);
+  }
+
+  @Override
+  public SortedMap<ArtifactDescriptor, Set<PluginClass>> getPlugins(
+    NamespaceId namespace, Id.Artifact artifactId, String pluginType) throws IOException, ArtifactNotFoundException {
+    return delegate.getPlugins(namespace, artifactId, pluginType);
+  }
+
+  @Override
+  public SortedMap<ArtifactDescriptor, PluginClass> getPlugins(
+    NamespaceId namespace, Id.Artifact artifactId, String pluginType, String pluginName,
+    Predicate<ArtifactId> pluginPredicate, int limit,
+    ArtifactSortOrder order) throws IOException, PluginNotExistsException, ArtifactNotFoundException {
+    return delegate.getPlugins(namespace, artifactId, pluginType, pluginName, pluginPredicate, limit, order);
+  }
+
+  @Override
+  public Map.Entry<ArtifactDescriptor, PluginClass> findPlugin(
+    NamespaceId namespace, ArtifactRange artifactRange, String pluginType, String pluginName,
+    PluginSelector selector) throws ArtifactNotFoundException, IOException, PluginNotExistsException {
+    return delegate.findPlugin(namespace, artifactRange, pluginType, pluginName, selector);
+  }
+
+  @Override
+  public ArtifactDetail addArtifact(Id.Artifact artifactId, File artifactFile) throws Exception {
+    return delegate.addArtifact(artifactId, artifactFile);
```

Comment:
This method should call `addArtifact(Id.Artifact artifactId, File artifactFile, @Nullable Set<ArtifactRange> parentArtifacts, @Nullable Set<PluginClass> additionalPlugins)` so that authorization checks are performed.